sympy/simplify/simplify.py

```diff
@@ -943,6 +943,76 @@ def _ratsimpmodprime(a, b, allsol, N=0, D=0):
     return (c*r.q)/(d*r.p)
 
 
+def rootofsimp(expr):
+    """
+    Reduce expressions containing ``RootOf`` types modulo their defining
+    polynomials.
+
+    A root `r` of a polynomial `p(x) = a_n x^n + ... + a_0` satisfies `p(r) =
+    0` by definition. Therefore, any power of `r` greater than `n` can be
+    rewritten in terms of smaller powers. In general, any polynomial or
+    rational expression of `r` can be reduced modulo its defining polynomial
+    `p`.
+
+    ``rootofsimp`` rewrites the expression as a numerator-denominator pair,
+    computes the modular inverse of the denominator, and returns the product of
+    this inverse with the numerator modulo the defining polynomial of the
+    root. This is done for each ``RootOf`` appearing in the expression.
+
+    Examples
+    ========
+
+    >>> from sympy import rootofsimp, RootOf
+    >>> from sympy.abc import x,y
+    >>> r = RootOf(x**4 - x - 1, 0)
+    >>> rootofsimp(r**4)
+    RootOf(x**4 - x - 1, 0) + 1
+    >>> rootofsimp(r**8)
+    2*RootOf(x**4 - x - 1, 0) + RootOf(x**4 - x - 1, 0)**2 + 1
+    >>> rootofsimp(1/r)
+    -1 + RootOf(x**4 - x - 1, 0)**3
+
+    """
+    from sympy.polys.rootoftools import RootOf
+    expr = sympify(expr)
+
+    # temporarily replace each RootOf with a dummy variable. this is necessary
+    # to preserve options such as "radicals=False" as well as for avoiding
+    # conflicts when the defining polynomial of multiple RootOfs use the same
+    # variable
+    rootofs = expr.find(RootOf)
+    dummies = [Dummy() for _ in rootofs]
+    transform = dict(zip(rootofs,dummies))
+    expr = expr.xreplace(transform)
+
+    # prepare the inverse dummy-variable-to-RootOfs transformation for the
+    # simpified expression. we create the inverse transform here so it can be
+    # applied to the expression if a 'NotInvertible' error occurs
+    transform = dict(zip(dummies,rootofs))
+
+    # simplify the expression in each root
+    for root,dummy in zip(rootofs,dummies):
+        try:
+            modulus = root.poly.xreplace({root.poly.gen:dummy})
+            numer,denom = cancel(expr).as_numer_denom()
+            denom = denom.as_poly(dummy).invert(modulus)
+            expr = numer.as_poly(dummy)*denom % modulus
+        except NotInvertible:
+            denom = denom.xreplace(transform).as_expr()
+            modulus = modulus.xreplace(transform).as_expr()
+            raise NotInvertible('The denominator, %s, is not invertible '
+                                'modulo %s'%(denom, modulus))
+        except:
```

Comment:
So...in some other test code I had a `TypeError` occur when `rootofsimp` is used on an expression that is not a polynomial expression of a `RootOf`. For example:

```python
from sympy import sin, RootOf, rootofsimp
from sympy.abc import x,y
a = RootOf(y**5 + y + 1, 0)
expr = sin(a**5)
rootofsimp(expr)
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-8dc45ed936b6> in <module>()
      3 a = RootOf(y**5 + 2*y + 1, 0)
      4 expr = sin(a**5)
----> 5 rootofsimp(expr)

/Users/cswiercz/sympy/sympy/simplify/simplify.py in rootofsimp(expr)
    997             numer,denom = cancel(expr).as_numer_denom()
    998             denom = denom.as_poly(dummy).invert(modulus)
--> 999             expr = numer.as_poly(dummy)*denom % modulus
   1000         except NotInvertible:
   1001             denom = denom.xreplace(transform).as_expr()

TypeError: unsupported operand type(s) for *: 'NoneType' and 'Poly'
```
I ran this using `pdb` and found out that Sympy is deciding to return a `None` on `numer.as_poly(dummy)`. Although, strangely, I can do the following outside of the `pdb` session:
```python
print expr.as_poly(a)
```
```
Poly(sin(RootOf(y**5 + 2*y + 1, 0)**5), RootOf(y**5 + 2*y + 1, 0), domain='EX')
```
Even more so, I can try to compute the modulus of this with respect to `a**5 + 2*a + 1` and Sympy returns the untouched expression instead of returning None. Not sure how best to proceed.